File: algorithms/213. House Robber II/solution.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if ( nums.size() > 1 ) {
            return max(robrange(nums, 0, nums.size() - 1), robrange(nums, 1, nums.size()));
        } else {
            return nums.size() ? nums[0] : 0;
        }
    }

    int robrange(vector<int> & nums, int start, int end) {
        int p1 = 0, p2 = 0, tmp;
        for ( int i = start; i < end; ++i ) {
            tmp = p2;
            p2 = max(p1 + nums[i], p2);
            p1 = tmp;
        }
        return p2;
    }
};

File: algorithms/213. House Robber II/solution.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        p0 = p1 = p2 = p3 = 0
        for i in xrange(0, len(nums) - 1):
            p0, p1 = p1, max(p0 + nums[i], p1)
        for i in xrange(1, len(nums)):
            p2, p3 = p3, max(p2 + nums[i], p3)
        return max(p1, p3) if len(nums) > 1 else sum(nums)
